ppl/zqd/recruiter/recruiter_test.go

```diff
@@ -0,0 +1,293 @@
+package recruiter
+
+import (
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func Register1(t *testing.T, addr string, nodename string, fp int, np int, np1 int, rp int) *WorkerPool {
+	wp := NewWorkerPool()
+	wp.Register(addr, nodename)
+	AssertPoolLen(t, wp, nodename, fp, np, np1, rp)
+	return wp
+}
+
+func AssertPoolLen(t *testing.T, wp *WorkerPool, name string, fp int, np int, np1 int, rp int) {
+	assert.Equal(t, fp, wp.LenFreePool(), "FreePool len=%d", fp)
+	require.Equal(t, np, wp.LenNodePool(), "NodePool len=%d", np)
+	assert.Equal(t, rp, wp.LenReservedPool(), "ReservedPool len=%d", rp)
+}
+
+func TestRegister1(t *testing.T) {
+	Register1(t, "a.b:5000", "n1", 1, 1, 1, 0)
+}
+
+func TestRegisterBadAddr(t *testing.T) {
+	wp := NewWorkerPool()
+	wp.Register("a.b;;5000", "n1")
+	AssertPoolLen(t, wp, "", 0, 0, -1, 0)
+}
+
+func TestRegisterBlankNode(t *testing.T) {
+	wp := NewWorkerPool()
+	wp.Register("a.b:5000", "")
+	AssertPoolLen(t, wp, "", 0, 0, -1, 0)
+}
+
+func TestRegisterTwice(t *testing.T) {
+	wp := Register1(t, "a.b:5000", "n1", 1, 1, 1, 0)
+	wp.Register("a.b:5000", "n1")
+	AssertPoolLen(t, wp, "n1", 1, 1, 1, 0)
+}
+
+func TestDeregister1(t *testing.T) {
+	wp := Register1(t, "a.b:5000", "n2", 1, 1, 1, 0)
+	wp.Deregister("a.b:5000")
+	AssertPoolLen(t, wp, "n2", 0, 0, 0, 0)
+}
+
+func TestRecruit1(t *testing.T) {
+	addr := "a.b:5000"
+	nodename := "n2"
+	wp := Register1(t, addr, nodename, 1, 1, 1, 0)
+	s, err := wp.Recruit(1)
+	require.Nil(t, err)
+	require.Len(t, s, 1)
+	AssertPoolLen(t, wp, nodename, 0, 0, 0, 1)
+	assert.Equal(t, addr, s[0].Addr)
+	// try recruiting 0 workers
+	_, err = wp.Recruit(0)
+	assert.EqualError(t, err, "recruit must request one or more workers: n=0")
+	// attempt to register reserved worker should be ignored
+	wp.Register(addr, nodename)
+	AssertPoolLen(t, wp, nodename, 0, 0, 0, 1)
+	wp.Unreserve(addr)
+	AssertPoolLen(t, wp, nodename, 0, 0, 0, 0)
+	// recruit with none available returns empty list
+	s, err = wp.Recruit(1)
+	assert.Nil(t, err)
+	assert.Len(t, s, 0)
+}
+
+func TestRegister10(t *testing.T) {
+	wp := NewWorkerPool()
+	var addr string
+	var nodename string
+	for i := 0; i < 5; i++ {
+		nodename = "n" + strconv.Itoa(i)
+		for j := 0; j < 2; j++ {
+			addr = nodename + ".x:" + strconv.Itoa(5000+j)
+			err := wp.Register(addr, nodename)
+			require.Nil(t, err)
+		}
+	}
+	AssertPoolLen(t, wp, nodename, 10, 5, 2, 0)
+	s, err := wp.Recruit(5)
+	require.Nil(t, err)
+	assert.Len(t, s, 5)
+	AssertPoolLen(t, wp, nodename, 5, 5, 1, 5)
+}
+
+func InitRectangle(t *testing.T, wp *WorkerPool, width int, height int) {
```

Comment:
'rectangle' and 'triangle' don't really help me understand what these do - can you change these names?

Reply:
I revised the names. BTW, the triangle is a set of nodes where every node has a different number of available workers.